src/Registration/cReg/NiftiImageData.cpp

```diff
@@ -1742,11 +1742,34 @@ void NiftiImageData<dataType>::dot(const DataContainer& a_x, void* ptr) const
 {
     const NiftiImageData<dataType>& x = dynamic_cast<const NiftiImageData<dataType>&>(a_x);
     ASSERT(_nifti_image->nvox == x._nifti_image->nvox, "dot operands size mismatch");
-    double s = 0.0;
-    for (unsigned i=0; i<this->_nifti_image->nvox; ++i)
-        s += double(_data[i] * x._data[i]);
+    float s = 0.0;
+    for (unsigned i = 0; i < this->_nifti_image->nvox; ++i)
+        s += _data[i] * x._data[i];
+    float* ptr_s = static_cast<float*>(ptr);
+    *ptr_s = s;
+}
+
+template<class dataType>
+void NiftiImageData<dataType>::sum(void* ptr) const
+{
+    float s = 0.0;
```

Comment:
again `float`. Far better to use `double` (if you don't need `complex`)

Reply:
Let me try to explain again in more detail.

I needed C interface to abstract `DataContainer` class. The problem is, it would not know the type of the data it works with, and C does not have templates. My solution to this problem is as follows. Firstly, I assume that data is either `float` or `comples_float_t` (i.e. `std::complex<float>`). Whenever a value needs to be computed (e.g. `dot` or `sum`), a zero complex float value is allocated, and the address of that complex value is passed to the interface function as `void*`. If the computed value is actually real, it will occupy the first 4 bytes of the allocated complex value. The Python interface, having received a complex value, checks the imaginary part and returns real value if it is zero and complex value otherwise.

Reply:
sure, but that's a C interface problem. I'm talking about the C++ interface. (such things as (unsafe) casting should be handled in the C interface in my opinion.

Reply:
C interface does not know the data type.

Reply:
I'm a bit lost. sorry. Where in C++ do you actually handle complex in these functions?

Reply:
ok. I also see that the C++ `DataContainer` is not templated in `DataType`, so it cannot have virtual functions that depend on `DataType`. A common problem with base classes, so this needs a longer term solution, but let's go ahead as-is now.

Possible longer term solutions:
- use the largest type in the C++ interface (i.e. std::complex<double>) as opposed to `void *`. Easy, but probably unexpected.
- use `std::any` (and if needed with a virtual function `std::any DataContainer::get_zero_data_element()` or so). More complicated and not very elegant.
- template `DataContainer` if possible
- put a class in between `RealDataContainer` and `ComplexDataContainer`, which have the intuitive signatures, as well as the generic ones, which do all the ugly casting, such that the derived classes don't need to.

Can you create an issue about this with some comments?